### PR TITLE
Feature #2321 tc_diag_changes

### LINKED ISSUE #141: Update the database schema to handle new columns and line types added to the MET-11.0.0-beta5 output.
## Describe the Enhancement ##
MET development version 11.0.0-beta5 changes the output from the TC-Pairs and TC-Stat tools in the 2 ways described below. This issue is to update the database schema accordingly to handle this new data for MET version 11.0.0.

  - PR dtcenter/MET#2346 for issue dtcenter/MET#2036 renames the TCMPR columns as follows:
  OLD: `NUM_MEMBERS TRACK_SPREAD DIST_MEAN MSLP_SPREAD MAX_WIND_SPREAD`
  NEW: `NUM_MEMBERS TRACK_SPREAD TRACK_STDEV MSLP_STDEV MAX_WIND_STDEV`
  - PR dtcenter/MET#2347 for issue dtcenter/MET#2321 adds 2 new columns to the TCDIAG line type as follows:
  OLD: `TOTAL INDEX DIAG_SOURCE N_DIAG`
  NEW: `TOTAL INDEX DIAG_SOURCE TRACK_SOURCE FIELD_SOURCE N_DIAG`

The new `TRACK_SOURCE` and `FIELD_SOURCE` columns are strings and should be handled the same as the `DIAG_SOURCE` string.

Please find updated sample output data in this tarfile:
   ftp.rap.ucar.edu/incoming/irap/met_help/NB20221116_develop_test_output.tar.gz

In particular see:
Updated TCMPR lines in `test_output/tc_pairs/al132020_CONSENSUS.tcst`.
Updated TCDIAG lines in `test_output/tc_pairs/al092022_20220926_DIAGNOSTICS.tcst`.

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
None needed.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required: @venitahagerty 
- [x] Select **scientist(s)** or **no scientist** required: none needed

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
None needed.

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues

Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### LINKED ISSUE #2321: Refine TCDIAG output from TC-Pairs as needed
## Describe the Enhancement ##
Pull request #2315 for issue #392 added a new TCDIAG line type to the output of TC-Pairs. That issue contained the majority of the work but some cleanup tasks remain. Please see below:

- [x] @jvigh write the **Storm Diagnostics** section of the tc-pairs.rst chapter.
- [ ] @jvigh (Unrelated to the MET repository) Consider adding METplus use case(s) to demonstrate the use of these diagnostics in TC-Pairs. When doing so, consider documenting details about these diagnostic data sources in the METplus [Verification Datasets Guide](https://metplus.readthedocs.io/en/latest/Verification_Datasets/index.html).
- [x] @JohnHalleyGotway update docs to remove references to the `-tcdiag` and `-lsdiag` command line options that were consolidated into a single `-diag` command line option.
- [ ] @JohnHalleyGotway investigate why tc-stat is relatively slow reading the tc-pairs output and try to speed it up.
- [ ] @KathrynNewman this is somewhat related. MET distributes an old version of the watch/warning file. Should we update that with each release?
- [x] @JohnHalleyGotway modify usage statement to indicate that LSDIAG_DEV is not currently supported but will be added in the future.
- [x] @jvigh will coordinate with CSU folks to figure out what we should name their diagnostics data source. Currently there's a namespace conflict between calling the CSU data "TCDIAG" as well as the output line type "TCDIAG" and the proposed new MET tool "TC-Diag".

Do NOT plan to do this for version 11.0.0:
- [ ] @jvigh and @KathrynNewman continue learning about the realtime and development LSDIAG data sources. Consider how realtime LSDIAG track information should be handled.
- [ ] @JohnHalleyGotway add support for the developmental LSDIAG data source which contains 20+ years of data in a single file.
   - [ ] Consider adding use_diag_track = TRUE/FALSE; config file option. If true, use the diag lat/lon location to define the track rather than matching to an existing ADECK track. This idea requires more thought. Would this be a benefit or just cause confusion We could add logic to bool TrackInfoArray::add_diag_data(DiagFile &diag_file, const StringArray &diag_name) for this. 

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
No impacts for this set of changes.

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by JohnHalleyGotway**
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes]**</br>
If **yes**, please describe:</br>

  - Switch to using diagnostic source names CIRA_DIAG_RT and SHIPS_DIAG_RT.
  - Adds TC-Pairs `diag_info_map` configuration option.
  - Move `diag_name` config option inside the larger `diag_info_map` option to define diagnostics metadata.
  - Rename `diag_convert_map.source` to `diag_convert_map.diag_source` for consistency, and support partial string matching in that context.
  - Remove `model=string` sub-option for the `-diag` command line option since it's replaced by `diag_info_map.match_to_track`.

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[Yes]**</br>
If **yes**, please describe:</br>

  - Add 2 new TCDIAG output columns for TRACK_SOURCE and FIELD_SOURCE.

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

  - Compiled locally and ran unit_tc_pairs.xml to confirm that they all run.
  - Will use GHA for this PR to update the input test dataset and check for unexpected diffs.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

  - @KathrynNewman please review configuration file changes, user's guide updates, and also test on seneca:
  ```/d1/projects/MET/MET_pull_requests/met-11.0.0/met-11.0.0-beta5/feature_2321/MET-feature_2321_tc_diag_changes```
  - @jvigh please review diagnostic settings in the TC-Pairs config file `diag_info_map` and confirm that these are correct.
  - @georgemccabe please review the code changes and configuration file updates.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**
  
  - Updates the TC-Pairs documentation about diagnostics by merging changes from @jvigh's branch.
  - Updates the description of TC-Pairs command line options, configuration options, and the TCDIAG line type.

- [x] Do these changes include sufficient testing updates? **[Yes]**

  - I kept the same test in place, but I reorganized the input diagnostics data files to match our updated naming conventions.

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

  - Adds 2 new output columns to the TCDIAG line type.
  - One TC-Pairs output file is modified (`al092022_20220926_DIAGNOSTICS.tcst`), but only by the addition of the `TRACK_SOURCE` and `FIELD_SOURCE` columns. The actual data values are NOT changed.

- [x] Please complete this pull request review by **[Wed 11/16/2022]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Development** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Review by georgemccabe — APPROVED** [review]
These code changes look good to me. I didn't see anything glaringly incorrect. I will use this info to update the METplus logic for PR dtcenter/METplus#1932

**Comment by JohnHalleyGotway** [comment]
I inspected the differences flagged in this [GHA run](https://github.com/dtcenter/METplus/discussions/1909#discussioncomment-4151411). And I see that there is a problem in the TCDIAG output written by the TC-Stat tool.

I'll work on fixing that now.

File: docs/Users_Guide/tc-pairs.rst [hunk 1]
```diff
@@ -258,52 +285,75 @@ ____________________
...
+
+The **diag_source** entry is one of the supported diagnostics data sources. The **track_source** entry is a string defining the ATCF ID of the track data used to define the locations at which diagnostics are computed. This string is written to the **TRACK_SOURCE** column of the TCDIAG output line. The **field_source** entry is a string describing the gridded data over which the diagnostic are computed. This string is written to the **FIELD_SOURCE** column of the TCDIAG output line type. The **match_to_track** entry specifies a comma-separated list of strings defining the ATCF ID(s) of the tracks to which these diagnostic values should be matched. The **diag_name** entry specifies a comma-separated list of strings for the tropical cyclone diagnostics of interest. If a non-zero list of diagnostic names is specified, only those diagnostics appearing in the list are written to the TCDIAG output line type. If defined as an empty list (default), all diagnostics found in the input are written to the TCDIAG output lines.
```
Comment (jvigh) [inline docs/Users_Guide/tc-pairs.rst]: Minor edit for clarity/accuracy:

Suggest changing: 
The **field_source** entry is a string describing the gridded data over which the diagnostic are computed.

to:
The **field_source** entry is a string describing the gridded model data from which the diagnostics are computed.

**Review by jvigh — APPROVED** [review]
I just had one minor edit for clarity in the documentation section.

File: docs/Users_Guide/tc-pairs.rst (hunk 1 above)
Comment (JohnHalleyGotway) [reply docs/Users_Guide/tc-pairs.rst]: Done. Fixed by [this commit](https://github.com/dtcenter/MET/commit/d7c04bfe7eba02d729c6feeb1ac39411a62c4d37). For one line edits, please consider "proposing" the exact change you'd like to see so I can just click the button to accept the change.

**Review by KathrynNewman — CHANGES_REQUESTED** [review]
The default settings for diag_source="SHIPS_DIAG_RT"; are:
track_source = "OFCL";
match_to_track=["OFCL"];

This results in warning messages that the track locations do not match the diagnostic location. This may result in confusion for a user running with the default settings. 

After discussion with @JohnHalleyGotway  - proposed changes are to only print a warning about locations NOT matching when diag_info_map.track_source = diag_info_map.match_to_track. When they do not match, print a DEBUG(4) log message.

The default track_source needs to change for SHIPS_DIAG_RT for this to work. John consulted Jonathan for advice on what to use for this string.

**Comment by JohnHalleyGotway** [comment]
Thanks for the feedback @KathrynNewman. I've made the requested changes in [this commit](https://github.com/dtcenter/MET/pull/2347/commits/274010493eb660e0af4d5ad9455f7e5b825c8319) and am re-requesting your review.

**Review by KathrynNewman — CHANGES_REQUESTED** [review]
Mark says: "The SHIPS model includes its own "tracker" - so, rather than _the SHIPS model itself_, I recommend stating _along a GFS forecast track defined by a SHIPS specific tracker_. (or _internal SHIPS tracker_?) @jvigh, do you agree? The disclaimer that the SHIPS tracker doesn't appear in the NHC adeck is important (good add!).

**Comment by JohnHalleyGotway** [comment]
Done:
* SHIPS_DIAG_RT: Real-time SHIPS diagnostics computed from a NWP model such as the Global Forecast System (GFS) model along a GFS forecast track defined by a SHIPS-specific tracker. Note that these SHIPS-derived forecast tracks do not appear in the NHC adeck data.

**Review by KathrynNewman — APPROVED** [review]
Looks good - thanks!

**Comment by jvigh** [comment]
I'm just coming online now after a busy day running kids around to
appointments.

Yes, I agree with what @Kathryn Newman ***@***.***> proposes. The
SHIPS_TRK is a reasonable choice. Or maybe even SHIPS_DEV_TRK.



Jonathan Vigh						
Project Scientist I, Joint Numerical Testbed	
Research Applications Laboratory (RAL)
National Center for Atmospheric Research (NCAR)
P.O. Box 3000                    tel: +1 (303) 497-8205
Boulder, CO 80307-3000   fax: +1 (303) 497-8171Jonathan's Staff Web
Page <http://staff.ral.ucar.edu/jvigh/> (CV, publications,
etc.)Tropical Cyclone Guidance Project
<http://hurricanes.ral.ucar.edu/> (real-time hurricane data)

Tropical Cyclone Data Project <https://verif.rap.ucar.edu/tcdata/>
(FLIGHT+, VDM+, TC-OBS datasets)

HurricaneRiskCalculator®️ <https://wxrisk.io/> (Now available --
Personalized hurricane wind hazard assessments for decision support)

During COVID-19, my working day is likely different from your working
day. Please do not feel obliged to reply to this email outside of your
normal working hours.

**Pull request merged**